File: android/app/src/main/java/world/abyy/droneinspection/Tiles.java
package world.abyy.droneinspection;

import java.util.ArrayList;
import java.util.List;

/**
 * Cutting a frame into pieces, and putting the findings back together.
 *
 * A port of web/js/tiles.js, kept in step with it.
 *
 * WHY A CROWD NEEDS THIS
 *     The detector's input is a fixed square. A 1920-wide frame squeezed into 640 of them
 *     turns a person forty pixels tall into thirteen, and a crowd shot comes back with five
 *     people rather than fifty: they are not being missed by a threshold, they are not
 *     being shown to the model at all.
 *
 *     Half that frame is 960 across, which the model squeezes by 1.5 rather than 3, and the
 *     same person arrives at twenty-four pixels. Same model, same weights, nearly twice the
 *     size on the thing being looked for.
 *
 * WHY TWO AND NOT SIX
 *     Six 320-pixel tiles was the old shape and it gave a person about the same pixels as
 *     this does. What it could not do is look at all of them at once: six tiles at the
 *     cycle rate meant a patch of ground got a look every sixth cycle, and everybody
 *     outside the current tile was coasting on a guess. Two tiles at 640 cost the same
 *     milliseconds and cover the whole frame every cycle.
 *
 *     Measured on four crowded VisDrone frames at the real cadence: 45% of the people
 *     reached and 1.92 numbers each, against 60% reached and 1.69 numbers each. Recall
 *     against tile count saturates at two - three, four and six tiles all measure 70% on a
 *     still frame - so the extra tiles were buying latency and nothing else.
 */
final class Tiles {

    static final int COLUMNS = 2;
    static final int ROWS = 1;
    static final int COUNT = COLUMNS * ROWS;

    /** Tiles overlap, so a person standing on a seam is whole in at least one of them. */
    static final float OVERLAP = 0.18f;

    /** Above this overlap, two boxes of the same class are the same thing seen twice. */
    static final float MERGE_IOU = 0.55f;

    /** Where a tile sits in a frame: x, y, width, height. */
    static float[] region(int index, int frameWidth, int frameHeight) {
        int wrapped = ((index % COUNT) + COUNT) % COUNT;
        int column = wrapped % COLUMNS;
        int row = wrapped / COLUMNS;

        float tileWidth = frameWidth / (float) COLUMNS;
        float tileHeight = frameHeight / (float) ROWS;
        float padX = tileWidth * OVERLAP;
        float padY = tileHeight * OVERLAP;

        float x = Math.max(0f, column * tileWidth - padX);
        float y = Math.max(0f, row * tileHeight - padY);
        return new float[]{
                x, y,
                Math.min(frameWidth - x, tileWidth + padX * 2),
                Math.min(frameHeight - y, tileHeight + padY * 2),
        };
    }

    /** Intersection over union, for deciding whether two boxes are one thing. */
    static float overlap(float[] a, float[] b) {
        float width = Math.min(a[2], b[2]) - Math.max(a[0], b[0]);
        float height = Math.min(a[3], b[3]) - Math.max(a[1], b[1]);
        if (width <= 0 || height <= 0) {
            return 0f;
        }
        float intersection = width * height;
        float union = (a[2] - a[0]) * (a[3] - a[1])
                + (b[2] - b[0]) * (b[3] - b[1]) - intersection;
        return union > 0 ? intersection / union : 0f;
    }

    /**
     * Fold a tile's findings into the frame's, dropping anything already there.
     *
     * A tile overlaps its neighbours and covers ground the full-frame pass also looked at,
     * so the same person arrives twice. Keeping both would draw two boxes on one person
     * and, far worse, count them twice. Where they are the same, the more confident box
     * wins, and that is usually the tile's: it saw three times as many of their pixels.
     */
    static List<Finding> merge(List<Finding> base, List<Finding> extra) {
        List<Finding> merged = new ArrayList<>(base);
        for (Finding candidate : extra) {
            boolean duplicate = false;
            for (int i = 0; i < merged.size(); i++) {
                Finding existing = merged.get(i);
                if (!existing.label.equals(candidate.label)) {
                    continue;
                }
                if (overlap(box(existing), box(candidate)) > MERGE_IOU) {
                    if (candidate.confidence > existing.confidence) {
                        merged.set(i, candidate);
                    }
                    duplicate = true;
                    break;
                }
            }
            if (!duplicate) {
                merged.add(candidate);
            }
        }
        return merged;
    }

    private static float[] box(Finding finding) {
        return new float[]{finding.x0, finding.y0, finding.x1, finding.y1};
    }

    private Tiles() {
    }
}

File: web/js/tiles.js
/**
 * Cutting a frame into pieces, and putting the findings back together.
 *
 * Pure arithmetic, in its own file so it can be tested without a browser. The worker does
 * the drawing and the detecting; the geometry and the merging are here, because they are
 * where the mistakes live. A tile mapped back one pixel wrong puts a box next to a person
 * instead of on them, and a merge that is too eager draws one box on two people while a
 * merge that is too shy counts one person twice.
 */

// Two tiles, not six. Recall against tile count saturates at two once the model's input is
// 640 - three, four and six all measure 70% on the same frames - and two is what lets every
// person be looked at every cycle instead of every sixth one. Tiles.java carries the
// numbers; keep the two in step.
export const TILE_COLUMNS = 2;
export const TILE_ROWS = 1;
export const TILE_COUNT = TILE_COLUMNS * TILE_ROWS;

/** Tiles overlap, so a person standing on a seam is whole in at least one of them. */
export const TILE_OVERLAP = 0.18;

/** Above this overlap, two boxes of the same class are the same thing seen twice. */
export const MERGE_IOU = 0.55;

/**
 * Where tile `index` sits in a frame of this size.
 *
 * @returns {{x:number, y:number, width:number, height:number}}
 */
export function tileRegion(index, frameWidth, frameHeight) {
  const wrapped = ((index % TILE_COUNT) + TILE_COUNT) % TILE_COUNT;
  const column = wrapped % TILE_COLUMNS;
  const row = Math.floor(wrapped / TILE_COLUMNS);

  const tileWidth = frameWidth / TILE_COLUMNS;
  const tileHeight = frameHeight / TILE_ROWS;
  const padX = tileWidth * TILE_OVERLAP;
  const padY = tileHeight * TILE_OVERLAP;

  const x = Math.max(0, column * tileWidth - padX);
  const y = Math.max(0, row * tileHeight - padY);
  return {
    x,
    y,
    width: Math.min(frameWidth - x, tileWidth + padX * 2),
    height: Math.min(frameHeight - y, tileHeight + padY * 2),
  };
}

/** Intersection over union, for deciding whether two boxes are one thing. */
export function overlap(a, b) {
  const width = Math.min(a[2], b[2]) - Math.max(a[0], b[0]);
  const height = Math.min(a[3], b[3]) - Math.max(a[1], b[1]);
  if (width <= 0 || height <= 0) return 0;
  const intersection = width * height;
  const union = (a[2] - a[0]) * (a[3] - a[1]) + (b[2] - b[0]) * (b[3] - b[1]) - intersection;
  return union > 0 ? intersection / union : 0;
}

/**
 * How much of a frame-pass box a tile box has to sit inside before the frame box is
 * treated as a blur over it rather than a rival for it.
 */
const CONTAINED = 0.7;

/**
 * Fold a tile's findings into the frame's.
 *
 * THE CASE THIS GETS WRONG IF IT IS DONE NAIVELY
 *     Two people standing close together come back from the full-frame pass as one wide
 *     box, because at that scale they are one blob of pixels. The tile, looking at the same
 *     ground three times larger, sees two. A merge that just asks "do these overlap?" then
 *     throws both tile boxes away in favour of the blob, and the app draws one box around
 *     two people and counts them as one. That was reported from a real flight and it is the
 *     exact opposite of what tiling is for.
 *
 *     So overlap alone does not decide it. A tile box that sits mostly *inside* a frame box
 *     is not a rival reading of the same thing, it is a closer reading of part of it, and
 *     the frame box is a blur over whatever the tile found. When that happens the frame box
 *     is dropped and every tile box in it is kept.
 *
 *     The two are treated as the same thing only when they are the same size and place -
 *     genuine double vision, one person seen by both passes - and then the more confident
 *     wins, which is usually the tile's.
 */
export function merge(base, extra, threshold = MERGE_IOU) {
  const merged = base.map((finding) => ({ ...finding }));
  const supersededByTile = new Set();

  for (const candidate of extra) {
    let duplicate = false;

    for (const existing of merged) {
      if (existing.label !== candidate.label) continue;
      if (supersededByTile.has(existing)) continue;

      // The same thing seen twice: same place, same size.
      if (overlap(existing.box, candidate.box) > threshold) {
        if (candidate.confidence > existing.confidence) {
          existing.box = candidate.box;
          existing.confidence = candidate.confidence;
        }
        duplicate = true;
        break;
      }

      // A closer look at part of it. The wide box was a blur over more than one thing, and
      // whatever the tile resolves inside it is the better answer.
      if (containment(candidate.box, existing.box) >= CONTAINED
        && area(candidate.box) < area(existing.box) * 0.75) {
        supersededByTile.add(existing);
      }
    }
    if (!duplicate) merged.push({ ...candidate });
  }

  return merged.filter((finding) => !supersededByTile.has(finding));
}

/** How much of `inner` lies inside `outer`, as a fraction of `inner`. */
export function containment(inner, outer) {
  const width = Math.min(inner[2], outer[2]) - Math.max(inner[0], outer[0]);
  const height = Math.min(inner[3], outer[3]) - Math.max(inner[1], outer[1]);
  if (width <= 0 || height <= 0) return 0;
  const size = area(inner);
  return size > 0 ? (width * height) / size : 0;
}

function area(box) {
  return (box[2] - box[0]) * (box[3] - box[1]);
}

/**
 * Map a box found inside a tile back to where it is in the whole frame.
 *
 * @param {number[]} box  in the tile's own drawn pixels
 * @param {{x:number, y:number}} region  where the tile was cut from
 * @param {number} scaleX  tile source pixels per drawn pixel
 * @param {number} scaleY
 */
export function toFrame(box, region, scaleX, scaleY) {
  return [
    region.x + box[0] * scaleX,
    region.y + box[1] * scaleY,
    region.x + box[2] * scaleX,
    region.y + box[3] * scaleY,
  ];
}

/** Every tile covers ground, and between them they cover all of it. */
export function coversFrame(frameWidth, frameHeight) {
  const covered = [];
  for (let i = 0; i < TILE_COUNT; i += 1) covered.push(tileRegion(i, frameWidth, frameHeight));
  return covered;
}
